modules/exadsBidAdapter.js

```diff
@@ -0,0 +1,465 @@
+      'id': validBidRequest.params.impressionId,
+      'bidfloor': validBidRequest.params.bidfloor,
+      'bidfloorcur': validBidRequest.params.bidfloorcur,
+      'video': {
+        'mimes': validBidRequest.params.stream.video.mimes,
+        'protocols': validBidRequest.params.stream.protocols,
+      },
+      'ext': validBidRequest.params.stream.ext
+    }];
+
+    bidRequestData.imp = imp;
+  }
+
+  utils.logInfo('PAYLOAD', bidRequestData, JSON.stringify(bidRequestData));
+  utils.logInfo('FINAL URL', endpointUrl);
+
+  return makeBidRequest(endpointUrl, bidRequestData);
+};
+
+function handleResRTB2Dot4(serverResponse, request) {
+  utils.logInfo('on handleResRTB_2_4 -> request:', request);
+  utils.logInfo('on handleResRTB_2_4 -> request json data:', JSON.parse(request.data));
+  utils.logInfo('on handleResRTB_2_4 -> serverResponse:', serverResponse);
+
+  let bidResponses = [];
+
+  if (serverResponse.hasOwnProperty('body') && serverResponse.body.hasOwnProperty('id')) {
+    utils.logInfo('Ad server response', serverResponse.body.id);
+    utils.logInfo('serverResponse.body.seatbid[0].bid[0]', serverResponse.body.seatbid[0].bid[0]);
+
+    const bidRq = JSON.parse(request.data);
+    const requestId = serverResponse.body.id;
+    const bidData = serverResponse.body.seatbid[0].bid[0];
+    const currency = serverResponse.body.cur;
+    const bidResponseAd = bidData.adm;
+    const pixelUrl = bidData.nurl.replace(/^http:\/\//i, 'https://');
+
+    const bannerInfo = utils.deepAccess(bidRq.imp[0], 'banner');
+    const nativeInfo = utils.deepAccess(bidRq.imp[0], 'native');
+    const videoInfo = utils.deepAccess(bidRq.imp[0], 'video');
+
+    let w; let h = 0;
+    let mediaType = '';
+    const native = {};
+
+    if (bannerInfo != null) {
+      w = bidRq.imp[0].banner.w;
+      h = bidRq.imp[0].banner.h;
+      mediaType = BANNER;
+    } else if (nativeInfo != null) {
+      const reqNative = JSON.parse(bidResponseAd);
+      reqNative.native.assets.forEach(asset => {
+        if (asset.img != null) {
+          const imgAsset = JSON.parse(bidRq.imp[0].native.request)
+            .native.assets.filter(asset => asset.img != null).map(asset => asset.img);
+          w = imgAsset[0].w;
+          h = imgAsset[0].h;
+          native.image = {
+            url: asset.img.url,
+            height: h,
+            width: w
+          }
+        } else if (asset.title != null) {
+          native.title = asset.title.text;
+        } else if (asset.data != null) {
+          native.body = asset.data.value;
+        } else {
+          utils.logWarn('bidResponse->', 'wrong asset type or null');
+        }
+      });
+      mediaType = NATIVE;
+    } else if (videoInfo != null) {
+      mediaType = VIDEO;
+    }
+
+    const bidResponse = {
+      requestId: requestId,
+      currency: currency,
+      ad: bidData.adm,
+      cpm: bidData.price,
+      creativeId: bidData.crid,
+      cid: bidData.cid,
+      width: w,
+      ttl: 360,
```

Comment:
where is this coming from?

Reply:
https://docs.prebid.org/dev-docs/publisher-api-reference/getBidResponses.html (search for `ttl`)
It is a mandatory property, If I don't add it, the winner bid won't be returned to the publisher web site and the creative won't be rendered.